heatmap/src/heatmap.rs

```diff
@@ -297,166 +344,153 @@ impl Heatmap {
         &self.summary
     }
 
-    // Internal function which handles reuse of older windows to store newer
-    /// values.
-    fn tick(&self, time: Instant) {
-        loop {
-            // quick check to see if this data point requires ticking forward
-            let next_tick = self.next_tick.load(Ordering::Relaxed);
-            if time < next_tick {
-                return;
-            } else {
-                // some expiration needs to happen, let's try to acquire the lock
-                //
-                // note: we use parking_lot mutex as it will not be poisoned by
-                // a thread panic while locked.
-                if let Some(_lock) = self.lock.try_lock() {
-                    // now that we have the lock, check that we still need to
-                    // tick forward
-                    if time < self.next_tick.load(Ordering::Relaxed) {
-                        return;
-                    }
-
-                    // calculate the index of the next current slice
-                    let current = self.current.load(Ordering::Relaxed);
-                    let mut next = current + 1;
-                    if next >= self.slices.len() {
-                        next -= self.slices.len();
-                    }
+    fn idx_delta(&self, idx: usize, delta: i64) -> usize {
+        (idx + (self.slices() as i64 + delta) as usize) % self.slices()
+    }
 
-                    // move current and next_tick forward
-                    self.current.store(next, Ordering::Relaxed);
-                    self.next_tick.fetch_add(self.resolution, Ordering::Relaxed);
+    // compute the current slice index based on `tick_at` and `tick_origin`
+    fn slice_idx(&self, tick_at: Instant) -> usize {
+        let ntick =
+            tick_at.duration_since(self.tick_origin).as_nanos() / self.resolution.as_nanos();
+        (ntick - 1) as usize % self.slices()
+    }
 
-                    // now we have a slice to subtract and clear from the summary
-                    // this is the histogram that is one ahead of our new current
-                    // position
-                    let mut to_clear = next + 1;
+    // Internal function which handles all the housekeeping tasks that come due
+    // as the clock advances- primarily updating the time windows covered by each
+    // individual histogram, and cleaning up the values stored in buckets when
+    // the histogram is assigned to handle a new time slice.
+    //
+    // It returns the `tick_at` value which indicates the upper bound of the
+    // current `span`, the index of the most recent `Histogram` slice, and by
+    // how many ticks the heatmap moved forward.
+    fn tick(&self, now: Instant) -> (Instant, usize, usize) {
+        loop {
+            loop {
```

Comment:
Unclear to me why you have a loop in a loop. I think this can be done with a single loop? If not, a comment would be great.

Reply:
This should be here. I think there used to be two loops (one after acquiring the mutex), but the refactoring essentially expanded the second loop all the way to the top so now it's redundant.